skimage/segmentation/trainable_segmentation.py

```diff
@@ -0,0 +1,302 @@
+
+    Parameters
+    ----------
+    clf : classifier object, optional
+        classifier object, exposing a ``fit`` and a ``predict`` method as in
+        scikit-learn's API, for example an instance of
+        ``RandomForestClassifier`` or ``LogisticRegression`` classifier.
+    features_func : function, optional
+        function computing features on all pixels of the image, to be passed
+        to the classifier. The output should be of shape
+        ``(m_features, *labels.shape)``. If None,
+        :func:`skimage.segmentation.multiscale_basic_features` is used.
+    downsample : int, optional
+        downsample the number of training points. Use downsample > 1 if you
+        built the training set by brushing through large areas but not all
+        points are needed to train efficiently the classifier. The training
+        time increases with the number of training points.
+
+    Methods
+    -------
+    fit
+    predict
+    """
+
+    def __init__(self, clf=None, features_func=None, downsample=10):
+        if clf is None:
+            try:
+                from sklearn.ensemble import RandomForestClassifier
+                self.clf = RandomForestClassifier(n_estimators=100, n_jobs=-1)
+            except ImportError:
+                raise ImportError(
+                    "Please install scikit-learn or pass a classifier instance"
+                    "to TrainableSegmenter."
+                        )
+        else:
+            self.clf = clf
+        self.features_func = features_func
+        self.downsample = downsample
+
+
+    def fit(self, image, labels):
+        """
+        Train classifier using partially labeled (annotated) image.
+
+        Parameters
+        ----------
+        image : ndarray
+            Input image, which can be grayscale or multichannel, and must have a
+            number of dimensions compatible with ``self.features_func``.
+        labels : ndarray of ints
+            Labeled array of shape compatible with ``image`` (same shape for a
+            single-channel image). Labels >= 1 correspond to the training set and
+            label 0 to unlabeled pixels to be segmented.
+        """
+        output, clf = fit_segmenter(image, labels, self.clf,
+                                    self.features_func, self.downsample)
+        self.segmented_image = output
+
+
+    def predict(self, image):
+        """
+        Segment new image using trained internal classifier.
+
+        Parameters
+        ----------
+        image : ndarray
+            Input image, which can be grayscale or multichannel, and must have a
+            number of dimensions compatible with ``self.features_func``.
+
+        Raises
+        ------
+        NotFittedError if ``self.clf`` has not been fitted yet (use ``self.fit``).
+        """
+        return predict_segmenter(image, self.clf, self.features_func)
+
+
+def fit_segmenter(image, labels, clf, features_func=None, downsample=10):
```

Comment:
@jni (and all), regarding the API, I'm thinking of changing this function to take an array of features first (the first dimension corresponding to the number of features), so that the features could be computed beforehand (useful for caching results or a warm restart of the classifier). What do you think?